x/nexus/keeper/keeper.go

```diff
@@ -68,26 +69,25 @@ func (k Keeper) GetParams(ctx sdk.Context) types.Params {
 	return p
 }
 
-// SetAddressValidator sets the nexus address validator. It will panic if called more than once
-func (k *Keeper) SetAddressValidator(validator types.AddressValidator) {
-	if k.addressValidator != nil {
-		panic("validator already set")
+// SetAddressValidators sets the nexus address validator. It will panic if called more than once
+func (k *Keeper) SetAddressValidators(validators *types.AddressValidators) {
+	if !validators.IsSealed() {
+		panic("address validator must be sealed")
 	}
 
-	k.addressValidator = validator
+	if k.addressValidators != nil {
+		panic("address validator already set")
+	}
 
-	// In order to avoid invalid or non-deterministic behavior, we seal the validator immediately
-	// to prevent additionals handlers from being registered after the keeper is initialized.
-	k.addressValidator.Seal()
+	k.addressValidators = validators
 }
 
-// getAddressValidator returns the nexus address validator. If not set, it returns a sealed empty validator
-func (k Keeper) getAddressValidator() types.AddressValidator {
-	if k.addressValidator == nil {
-		k.SetAddressValidator(types.NewAddressValidator())
+func (k Keeper) getAddressValidator(module string) (exported.AddressValidator, error) {
+	if k.addressValidators == nil {
```

Comment:
why create an empty instance instead of return error if not intialized?

Reply:
it's not necessarily wrong to create a keeper without validators and this makes it easier to use. I don't have strong opinions about it either way, therefore I kept with Sammy's original implementation

Reply:
The previous function does not return error, so it returns a default one and do the check.

With the refactor, we have the ability to distinguish errors  between validators not initialized and module not registered. We can hide also this kind of info from the caller, no strong option. `k.addressValidators.GetAddressValidator(module)` will still return error on the default validators anyway.